### P003 ###


##### *1. O que é uma exceção em Java e qual é o propósito de usá-las em programas?* #####

###### Quando eventos anormais acontece em um codigo por exemplo uma entrada de codigo com o tipo diferente do que se é exigido, divisões por zero, entres outras... o programa procura um bloco catch para tratar esse problema, quando não é encontrado pode resultar em um erro, parando o programa. Para entradas inesperadas pelo usuario e geralmente inputs não logicos para o sistema ou o objetivo da aplicação se usa em java o sistema de tratamento de erros  da linguagem, que é tambem baseado em classe e possui hiêrarquia 'Throwable' que é a base para todas as exceções e possuir duas sub-classes 'Error" e 'Exception', sendo a primeira para erros que geralmente o programa não conseguir solucionar e a segunda para exceções que podem ser tratadas. #######


##### *2. Pesquise sobre a diferença entre exceções verificadas e não verificadas em Java. Dê exemplos de cada uma.* #####


###### Exceções verificadas o compilador obriga vc a tratar os error com 'try' e 'catch' ou declarar que seu metodo pode lançar essa exceção com a palavra-chave throws . ######

~~~ java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class ExemploCheckedException {

    public static void main(String[] args) {
        try {
            // Código que pode lançar uma exceção verificada
            lerArquivo("arquivo.txt");
        } catch (IOException e) {
            // Tratamento da exceção
            System.err.println("Erro de E/S: " + e.getMessage());
        }
    }

    public static void lerArquivo(String nomeArquivo) throws IOException {
        // Código que pode lançar IOException
        BufferedReader reader = new BufferedReader(new FileReader(nomeArquivo));
        // Restante do código...
    }
}

~~~


###### Exceções nao verificadas o compilador não obriga a você tratar o erro explicitamente. Em geral esse erro é vem da logica no codigo, sendo usada bastante para logica de negocios.



~~~ java
import java.util.Scanner;

public class CalculadoraDivisora {
	private int x;
	private int y;



void Dividir()
{
try {
	Scanner sc = new Scanner(System.in);
	
	System.out.println("Digite um numero do tipo inteiro para a divisão");
	x = sc.nextInt();
	System.out.println("Digite outro numero do tipo inteiro para a divisão");
	y = sc.nextInt();
	

	
	int dividindo = x / y;
	System.out.println("A divisão de " + x + " / " + y + " = " + dividindo );	
	}catch (Exception ex) {
	System.out.println("Error - valor digitado não é do tipo inteiro!");
	}		
	//exception esta agindo quando o valor digitado e diferente de inteiro
}

public static void main(String[] args) 
{
	CalculadoraDivisora divisora;
	divisora = new CalculadoraDivisora();
	
	divisora.Dividir();
	
	
	}

}

~~~


##### 3. *Como você pode lidar com exceções em Java? Quais são as palavras-chave e as práticas comuns para tratamento de exceções?* #####

###### Para lidar com exceções em **Java** existe um tratamento que se usa os blocos 'try', 'catch' e quando adequado o bloco 'finally'. Com o bloco 'try' envolve o codigo, o bloco 'catch' captura e trata as exceçoes e o bloco 'finally' é usado para garantir que determinadas ações sejam realizadas, independentemente de ocorrer uma exceção ou não. Isso é crucial para manter a consistência, evitar vazamentos de recursos e realizar ações finais necessárias para a integridade do programa como gravaçoes de logs ou restaurar a consitencia do programa entre outras utilidades. tambem existe palavras especificas para cada tipo de exceção e uma palavra-chave quando pode ocorrer uma delas que é 'throw'.  #######


#####  *4. O que é o bloco "try-catch" em Java? Como ele funciona e por que é importante usá-lo ao lidar com exceções?  * #####

###### O bloco "try-cath" é utiizado para tratar exceções o try por exemplo contém o código que pode gerar exceções. Este é o bloco no qual você monitora a ocorrência de exceções. Ja p catch fornece blocos para tratar exceções específicas. Cada bloco catch especifica o tipo de exceção que pode ser tratado e fornece um bloco de código para executar em caso de ocorrência dessa exceção. A importancia de usa-los é por diversos motivos sendo evitar erros inesperados, programação defensiva e na manutênção da execução do programa. ######

##### *5. Quando é apropriado criar suas próprias exceções personalizadas em Java e como você pode fazer isso? Dê um exemplo de quando e por que você criaria uma exceção personalizada.* #####


###### É apropriado quando se esta fazendo uma manutenção no codigo e precisa tratar problemas especificos e para diversas outras situações especificas que necessitam de comentarios de algum modo ######

##### **Calculadora de Divisão** #####

~~~ java
import java.util.Scanner;

class DivisionByZeroException extends Exception {

    // Construtores
    public DivisionByZeroException() {
        super("Divisão por zero não permitida.");
    }

    public DivisionByZeroException(int x, int y) {
        super("Divisão por zero: " + x + " / " + y);
    }
}


public class Calculadora1 {
	private int x;
	private int y;
	private int escolha;
	private float opcao;
	private float y1;
	private float x1;
	
	
	
	public 
	void somando()			//metodo para somar.
	{
		int somar = x + y;
		System.out.println("A Soma de " + x + "+" + y + " = "+ somar);
	}
	
	
	void subtraindo()		//metodo para subtrair.
	{
		int subtrair = x - y;
		System.out.println("A Subtracao de " + x +  "-" + y + " = " + subtrair);
	}
	
	void multiplicando()	//metodo para multiplicar.
	{
		int multiplicar = x * y;
		System.out.println("A Multiplicacao de " + x + "*" + y + " = " + multiplicar + "\n");
	}
	
	void dividindo () throws DivisionByZeroException	//metodo para dividir.
	{
		if (y < 1) {
			throw new DivisionByZeroException(x, y);
		}
		int dividir = x / y;
		System.out.println("A Divisao de "+ x + "/" + y +" = " + dividir);
	}
	
	void dividindoDecimal()	throws DivisionByZeroException	//metodo para dividir decimal.
	{
	
		if (y1 == 0) {
			throw new DivisionByZeroException();
		}
		float dividir =  x1 / y1;
		System.out.println("A Divisao de "+ x1 + "/" + y1 +" = " + dividir);
		
	
	
		
	}
	
	
	//metodo para interface e saida.
	void calculando() throws DivisionByZeroException
	{
		//um for para criar o loop da lista
		for(int i = 0; i <= 5;i++ ) {	
			Scanner sc = new Scanner(System.in);
			
			System.out.println("1. Divisão numeros decimais\n2. Calculadora Numeros inteiro");
			opcao = sc.nextFloat();
			if(opcao == 1)
			{
				System.out.println("Digite o Dividendo decimal");
				x1 = sc.nextFloat();
				System.out.println("Digite o Divisor decimal");
				y1 = sc.nextFloat();
				
				try {
					dividindoDecimal();
				}catch(DivisionByZeroException e) {
					System.out.println("Erro " + e.getMessage());
				}
				
			}
			System.out.println("Digite o Primeiro numero: ");
			
			x = sc.nextInt();
			System.out.println("Digite o numero o Segundo numero: ");
			y = sc.nextInt();
			
			System.out.println("Digite 0. Sair\nDigite 1. Somar\nDigite 2. Subtrair\nDigite 3. Multiplicar\nDigite 4. Dividir\n");
			escolha = sc.nextInt();
			
			// if para criar condiçoes a qual o usuario pode optar.
			if(escolha == 1) 
			{
				somando();
			}
			if(escolha == 2) {
				subtraindo();
			}
			if(escolha == 3) {
				multiplicando();
			}
			if(escolha == 4) {
				try {
				dividindo();
				}catch(DivisionByZeroException e) {
					System.out.println("Erro " + e.getMessage());
			}
			}
			if(escolha == 5) {
			try {
				dividindoDecimal();
			}catch(DivisionByZeroException e) {
				System.out.println("Erro " + e.getMessage());
			}
			}
			else if(escolha > 5)
			{
				System.out.println("Opção invalida. Tente novamente!");
			}
			else if(escolha == 0)
			{
				break;
			}
		}
	}
	



public int getX() {
		return x;
	}

	public void setX(int x) {
		this.x = x;
	}

	public int getY() {
		return y;
	}

	public float getOpcao() {
		return opcao;
	}


	public void setOpcao(float opcao) {
		this.opcao = opcao;
	}


	public float getY1() {
		return y1;
	}


	public void setY1(float y1) {
		this.y1 = y1;
	}


	public float getX1() {
		return x1;
	}


	public void setX1(float x1) {
		this.x1 = x1;
	}


	public void setY(int y) {
		this.y = y;
	}

	public int getEscolha() {
		return escolha;
	}

	public void setEscolha(int escolha) {
		this.escolha = escolha;
	}

	
public static void main(String[] args) throws DivisionByZeroException 
	{	
		Calculadora1 c;
		c = new Calculadora1();
		
		
		c.calculando();
		//Impressão do codigo.
	
	}	
	
}

~~~
##### Referências #####

###### 11 Erros que desenvolvedores Java cometem quando usam Exceptions: https://www.oracle.com/br/technical-resources/article/java/erros-java-exceptions.html ######


###### Tratamento de exceções em Java: https://www.universidadejava.com.br/java/java-excecoes/ ######



###### Como criar uma Exception/Exceção customizada em Java: https://pt.stackoverflow.com/questions/71670/como-criar-uma-exception-exce%C3%A7%C3%A3o-customizada-em-java ######


###### Curso de Java 47: Exceptions: try, catch: https://www.youtube.com/watch?v=ld2C4GcAtsg ######


###### Java - Aula 19 - O que são checked e unchecked exceptions?: https://www.youtube.com/watch?v=U2_qUtQvtGk ######



























